onadata/apps/api/viewsets/data_viewset.py

```diff
@@ -411,6 +413,77 @@ class DataViewSet(AnonymousUserPublicFormsMixin, ModelViewSet):
 
     queryset = XForm.objects.all()
 
+    def bulk_delete(self, request, *args, **kwargs):
+        """
+        Bulk delete instances
+        """
+        xform = self.__validate_permission_on_bulk_action(request,
+                                                          'change_xform')
+        payload = self.__get_payload(request)
+
+        mongo_query = ParsedInstance.get_base_query(xform.user.username,
+                                                    xform.id_string)
+        postgres_query = {'xform_id': xform.id}
+
+        instances_ids = self.__get_instances_ids(payload, mongo_query)
+
+        if instances_ids is not None:
+            # Narrow down queries with list of ids.
+            postgres_query.update({'id__in': instances_ids})
+            mongo_query.update({'_id': {'$in': instances_ids}})
+
+        # Delete Postgres & Mongo
+        updated_records_count = Instance.objects.filter(**postgres_query).count()
+
+        # Since Django 1.9, `.delete()` returns an dict with number of rows
+        # deleted per object.
+        # FixMe remove `.count()` query and use that dict instance
+        Instance.objects.filter(**postgres_query).delete()
+        ParsedInstance.bulk_delete(mongo_query)
+        return Response({
+            'detail': _('{} submissions have been deleted').format(
+                updated_records_count)
+        }, status.HTTP_200_OK)
+
+    def bulk_validation_status(self, request, *args, **kwargs):
+
+        xform = self.__validate_permission_on_bulk_action(request,
+                                                          'validate_xform')
+        payload = self.__get_payload(request)
+
+        if request.data.get('reset'):
+            new_validation_status_uid = {}
+        else:
+            new_validation_status_uid = payload.get('validation_status.uid')
+
+        if new_validation_status_uid is None:
+            raise ParseError(_('No `validation_status.uid` provided'))
+
+        # Create new validation_status object
+        new_validation_status = get_validation_status(
+            new_validation_status_uid, xform, request.user.username)
+
+        mongo_query = ParsedInstance.get_base_query(xform.user.username,
+                                                    xform.id_string)
+        postgres_query = {'xform_id': xform.id}
+
+        instances_ids = self.__get_instances_ids(payload, mongo_query)
```

Comment:
Idea:

* Have `__get_instances_ids()` take the `XForm` as an argument instead of the (base) `mongo_query`;
* Put the `ParsedInstance.get_base_query()` call in that method;
* Put the narrowing logic (from `if instances_ids is not None` below) also inside that method;
* Have the method return both the Mongo and Postgres queries (or pass them as `dict`s and have the function mutate them)
* Maybe rename the method now that it does something different :wink:

Reply:
Good idea ;-)